### DISCUSSION TITLE
m:n relation assignment with one of 2 foreign keys

### DISCUSSION THREAD

**Discussion by Langstra** [mikro-orm/mikro-orm#3654] [Ideas]
I was looking at the tests from GH3599 and saw that when adding new groups to a member, you still have to specify the member id.  To illustrate the example a little better these are the entities:
| **Member**                      | **GroupMember**                  | **Group**                        |
|---------------------------------|----------------------------------|----------------------------------|
| id: number                      | member: WrappedReference<Member> | id: number                       |
| groups: Collection<GroupMember> | group: WrappedReference<Group>   | members: Collection<GroupMember> |

When assigning a new group to a member you have to do it like this:
```ts
member.groups.set([orm.em.create(GroupMember, { group: group1, member })]);
```
It seems unnecessary to me to specify `member` again. We already know that `member` is going to be a value of the second primary key, since we are setting the groups on a specific member. However, since we are using `em.create` which create a new instance of `GroupMember` first, before setting it, I can understand this.

Then we have `em.assign`. We assign new groups to a member like this:
```ts
orm.em.assign(member, {
  groups: [
    {
      group: group1.id,
      member,
      accessLevel: 1,
    },
  ],
});
```
This definitely seems weird to me. Why do I need to specify `member` again, I am already assigning this group to `member. Sure, underneath it probably calls the same methods are `Collection.set`, but it seems unnecessary to me.

My idea is to have the assignment we without the relation property we are assigning it to.
```ts
member.groups.set([orm.em.create(GroupMember, { group: group1 })]);
// and also like this
orm.em.assign(member, {
  groups: [
    {
      group: group1.id,
      accessLevel: 1,
    },
  ],
});
```

Let me know what you think.

**Comment by B4nan** [comment]
> This definitely seems weird to me. Why do I need to specify member again, I am already assigning this group to member. Sure, underneath it probably calls the same methods are Collection.set`, but it seems unnecessary to me.

What is weird about this? Your `assign` call is replacing the whole collection items. You don't provide entity instances, but POJOs. To understand to what entity the POJO maps, we need to know its PK. Your entity has composite PK, we need to know all its parts to be able to construct the entity and check the identity map if it exists or not.

> member.groups.set([orm.em.create(GroupMember, { group: group1 })]);

This is a no-go to me, the entity instance needs to be valid at all time, and here you'd literally like to create an incomplete one. The `member.groups.set` call, which is interpreted _after_ the `em.create` call, is not relevant here. 

I am not against the latter, but I am not interested in implementing it either (so nope, don't create a new issue for that :]). Feel free to give that a try if you think it's that important, but hear me out, it will be very tricky. Propagation is implemented on the entity level, where we know everything about the metadata. If you want this to work, you'd probably have to implement it on POJOs too. The current implementation is simply not developed to support this, it actually proactively validates you don't do this.

> I was looking at the tests from GH3599 and saw that when adding new groups to a member, you still have to specify the member id

I feel like you are still missing one important thing. Those two lines are doing the same:

```ts
member.groups.set([orm.em.create(GroupMember, { group: group1, member })]);

// this line on its own is the same, as the `GroupMember.member` value is propagated to `member.groups` automatically
orm.em.create(GroupMember, { group: group1, member });
```

(or not exactly the same, the latter is adding to the `member.groups`, not replacing the state)

**Comment by B4nan** [comment]
I am starting to change my mind around the second idea - both in `em.create` and `em.assign`, it would be indeed nice to somehow automatically propagate the parent entity to the payload. 

What changed it was one test where I ended up with pretty much the same problem - I tried to introduce the reference wrapper to a project, the entity in question has a constructor requiring a `user` property, and now I want to do `wrap(user).toReference()` in the constructor to be able to assign it, and immediately see failures caused in seeding that uses `em.create` - the propagation works, but the entity instance gets created without the `user` parameter, so now the `wrap(user).toReference()` call fails as `user` is `undefined`.

I would gradually like to push the reference wrapper more, maybe even thinking of making it the default way in v6. And for that things like the above need to work.

(my stance about the `member.groups.set([orm.em.create(GroupMember, { group: group1 })]);` remains the same)
